diff --git a/jdk/src/share/native/java/net/net_util.c b/jdk/src/share/native/java/net/net_util.c
index b1b8223a63..b5db51eeef 100644
--- a/jdk/src/share/native/java/net/net_util.c
+++ b/jdk/src/share/native/java/net/net_util.c
@@ -38,7 +38,7 @@ JNIEXPORT jint JNICALL ipv6_available()
 }
 
 JNIEXPORT jint JNICALL
-JNI_OnLoad(JavaVM *vm, void *reserved)
+__libnet_JNI_OnLoad(JavaVM *vm, void *reserved)
 {
     JNIEnv *env;
     jclass iCls;
diff --git a/jdk/src/solaris/native/java/lang/java_props_md.c b/jdk/src/solaris/native/java/lang/java_props_md.c
index b03bd6ac8f..4114a25983 100644
--- a/jdk/src/solaris/native/java/lang/java_props_md.c
+++ b/jdk/src/solaris/native/java/lang/java_props_md.c
@@ -459,6 +459,10 @@ GetJavaProperties(JNIEnv *env)
     {
 #ifdef MACOSX
         setOSNameAndVersion(&sprops);
+#else
+#ifdef __EMSCRIPTEN__
+		sprops.os_name = strdup("Linux");
+		sprops.os_version = strdup("7.0.0-emscripten");
 #else
         struct utsname name;
         uname(&name);
@@ -477,6 +481,7 @@ GetJavaProperties(JNIEnv *env)
 #else
         sprops.os_version = strdup(name.release);
 #endif /* _AIX   */
+#endif /* __EMSCRIPTEN__ */
 #endif /* MACOSX */
 
         sprops.os_arch = ARCHPROPNAME;
@@ -494,34 +499,13 @@ GetJavaProperties(JNIEnv *env)
     sprops.sun_arch_abi = JDK_ARCH_ABI_PROP_NAME;
 #endif
 
-    /* Determine the language, country, variant, and encoding from the host,
-     * and store these in the user.language, user.country, user.variant and
-     * file.encoding system properties. */
-    setlocale(LC_ALL, "");
-    if (ParseLocale(env, LC_CTYPE,
-                    &(sprops.format_language),
-                    &(sprops.format_script),
-                    &(sprops.format_country),
-                    &(sprops.format_variant),
-                    &(sprops.encoding))) {
-        ParseLocale(env, LC_MESSAGES,
-                    &(sprops.language),
-                    &(sprops.script),
-                    &(sprops.country),
-                    &(sprops.variant),
-                    NULL);
-    } else {
-        sprops.language = "en";
-        sprops.encoding = "ISO8859-1";
-    }
+	sprops.language = "en";
+	sprops.encoding = "ISO8859-1";
     sprops.display_language = sprops.language;
     sprops.display_script = sprops.script;
     sprops.display_country = sprops.country;
     sprops.display_variant = sprops.variant;
 
-    /* ParseLocale failed with OOME */
-    JNU_CHECK_EXCEPTION_RETURN(env, NULL);
-
 #ifdef MACOSX
     sprops.sun_jnu_encoding = "UTF-8";
 #else
diff --git a/jdk/src/solaris/native/sun/nio/ch/FileDispatcherImpl.c b/jdk/src/solaris/native/sun/nio/ch/FileDispatcherImpl.c
index 9b3ff8194e..a1edb76bc8 100644
--- a/jdk/src/solaris/native/sun/nio/ch/FileDispatcherImpl.c
+++ b/jdk/src/solaris/native/sun/nio/ch/FileDispatcherImpl.c
@@ -72,7 +72,7 @@ Java_sun_nio_ch_FileDispatcherImpl_init(JNIEnv *env, jclass cl)
 {
     int sp[2];
     if (socketpair(PF_UNIX, SOCK_STREAM, 0, sp) < 0) {
-        JNU_ThrowIOExceptionWithLastError(env, "socketpair failed");
+		printf("[IKVM] socketpair failed\n");
         return;
     }
     preCloseFD = sp[0];
@@ -225,6 +225,16 @@ Java_sun_nio_ch_FileDispatcherImpl_lock0(JNIEnv *env, jobject this, jobject fdo,
                                       jboolean block, jlong pos, jlong size,
                                       jboolean shared)
 {
+#if defined(__EMSCRIPTEN__)
+    (void)env;
+    (void)this;
+    (void)fdo;
+    (void)block;
+    (void)pos;
+    (void)size;
+    (void)shared;
+    return sun_nio_ch_FileDispatcherImpl_LOCKED;
+#else
     jint fd = fdval(env, fdo);
     jint lockResult = 0;
     int cmd = 0;
@@ -256,12 +266,21 @@ Java_sun_nio_ch_FileDispatcherImpl_lock0(JNIEnv *env, jobject this, jobject fdo,
         JNU_ThrowIOExceptionWithLastError(env, "Lock failed");
     }
     return 0;
+#endif
 }
 
 JNIEXPORT void JNICALL
 Java_sun_nio_ch_FileDispatcherImpl_release0(JNIEnv *env, jobject this,
                                          jobject fdo, jlong pos, jlong size)
 {
+#if defined(__EMSCRIPTEN__)
+    (void)env;
+    (void)this;
+    (void)fdo;
+    (void)pos;
+    (void)size;
+    return;
+#else
     jint fd = fdval(env, fdo);
     jint lockResult = 0;
     struct flock64 fl;
@@ -279,6 +298,7 @@ Java_sun_nio_ch_FileDispatcherImpl_release0(JNIEnv *env, jobject this,
     if (lockResult < 0) {
         JNU_ThrowIOExceptionWithLastError(env, "Release failed");
     }
+#endif
 }
 
 
