diff --git a/configure.ac b/configure.ac
index dcfc7b2..5ed7079 100644
--- a/configure.ac
+++ b/configure.ac
@@ -225,7 +225,10 @@ AC_ARG_ENABLE(pax_emutramp,
       [Define this if you want to enable pax emulated trampolines (experimental)])
   fi)
 
-LT_SYS_SYMBOL_USCORE
+m4_ifdef([LT_SYS_SYMBOL_USCORE],
+  [LT_SYS_SYMBOL_USCORE],
+  [sys_symbol_underscore=no
+   AC_SUBST([sys_symbol_underscore])])
 if test "x$sys_symbol_underscore" = xyes; then
     AC_DEFINE(SYMBOL_UNDERSCORE, 1, [Define if symbols are underscored.])
 fi
diff --git a/src/wasm/ffi.c b/src/wasm/ffi.c
index 486ffa7..7bf6856 100644
--- a/src/wasm/ffi.c
+++ b/src/wasm/ffi.c
@@ -156,8 +156,6 @@ _Static_assert(FFI_BAD_TYPEDEF_MACRO == FFI_BAD_TYPEDEF, "FFI_BAD_TYPEDEF must b
 ffi_status FFI_HIDDEN
 ffi_prep_cif_machdep(ffi_cif *cif)
 {
-  if (cif->abi != FFI_EMSCRIPTEN_ABI)
-    return FFI_BAD_ABI;
   // This is called after ffi_prep_cif_machdep_var so we need to avoid
   // overwriting cif->nfixedargs.
   if (!(cif->flags & VARARGS_FLAG))
@@ -791,8 +789,6 @@ ffi_prep_closure_loc_js,
 ffi_status ffi_prep_closure_loc(ffi_closure *closure, ffi_cif *cif,
                                 void (*fun)(ffi_cif *, void *, void **, void *),
                                 void *user_data, void *codeloc) {
-  if (cif->abi != FFI_EMSCRIPTEN_ABI)
-    return FFI_BAD_ABI;
   return ffi_prep_closure_loc_js(closure, cif, (void *)fun, user_data,
                                      codeloc);
 }
